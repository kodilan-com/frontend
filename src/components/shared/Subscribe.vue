<script>
import { mapActions } from 'vuex';

export default {
  props: {
    fixed: {
      type: Boolean,
      required: false,
      default: false,
    },
  },
  data() {
    return {
      frequency: 'weekly',
      name: '',
      email: '',
      isClosed: false,
      frequencyModel: {
        value: 'weekly',
        text: 'haftalık',
      },
      frequencyOptions: [
        {
          value: 'weekly',
          text: 'haftalık',
        },
        {
          value: 'monthly',
          text: 'aylık',
        },
      ],
    };
  },
  computed: {
    isVisible() {
      if (!this.fixed) {
        return true;
      }

      return !this.isClosed && !localStorage.getItem('subscribeClosed');
    },
  },
  methods: {
    ...mapActions(['subscribe']),
    handleSubscribe() {
      this.frequency = this.frequencyModel.value;

      const { frequency, name, email } = this;

      this.subscribe({
        frequency,
        name,
        email,
      })
        .then(() => {
          this.showDialog('Email listesine kaydınız gerçekleştirildi.');
          this.close();
        })
        .catch((e) => {
          const details = Object.values(e.response.data.errors || [])
            .map(r => `<li>${r[0]}</li>`);
          this.showDialog('Hata: Kaydınız gerçekleştirilemedi.', `<ul>${details || ''}</ul>`);
        });
    },
    showDialog(title, text, buttons = [{ title: 'Kapat' }]) {
      this.$modal.show('dialog', {
        title,
        text,
        buttons,
      });
    },
    close() {
      this.isClosed = true;
      localStorage.setItem('subscribeClosed', true);
    },
  },
};
</script>

<template>
  <div
    v-if="isVisible"
    :class="{ fixed: fixed }"
    class="subscribe-widget"
  >
    <form @submit.prevent="handleSubscribe">
      <span>Güncel iş ilanlarını</span>
      <div class="subscribe--select">
        <multiselect
          v-model="frequencyModel"
          :options="frequencyOptions"
          label="text"
          :searchable="false"
          :close-on-select="true"
        />
      </div>
      <span>olarak almak için</span>
      <input type="text" v-model="name" placeholder="İsim soyisim" class="username">
      <input type="text" v-model="email" placeholder="Email">
      <button class="button" @click="handleSubscribe">
        Abone ol!
      </button>
      <span
        v-if="fixed"
        class="close"
        @click="close"
      >
        Bir daha gösterme :(
      </span>
    </form>
  </div>
</template>

<style lang="scss">
  .subscribe-widget {
    padding: 10px;
    box-sizing: border-box;
    font-size: 16px;
    text-align: center;

  .subscribe--select {
    text-align: left;
  }

  &.fixed {
    position: fixed;
    bottom: 0;
    left: 0;
    margin: 0;
    z-index: 10000;
    width: 100%;
    height: 70px;
    color: #fff;
    background: #231F20;
    border-top: 1px solid #474646;
    line-height: 50px;

      input[type=text] {
        margin-right: 6px;
        width: 120px;
        display: inline;

        &.username {
          margin-left: 6px;
        }
      }

      select,
      .subscribe--select {
        display: inline-block;
        width: 100px;
        margin: 0 6px;
        color: #202020;
      }

      .close {
        display: block;
        color: #5c5c5c;
        font-size: 12px;
        position: absolute;
        top: 10px;
        right: 10px;
        line-height: 12px;
        cursor: pointer;

        &:hover {
          color: #ccc;
        }
      }
    }

    input[type=text] {
      padding: 11px 10px;
    }

    select {
      height: 42px;
    }

    button {
      &:hover {
        background-color: #26ae61;
      }
    }
  }

  @media only screen and (max-width: 768px) {
    #footer {
      padding-bottom: 10px !important;
    }

    .subscribe-widget.fixed {
      display: none;
    }
  }

  @media only screen and (max-width: 998px) {
    .subscribe-widget.fixed {
      height: 80px;
    }
    .subscribe-widget form {
      margin-top: 10px;
    }
    .subscribe-widget.fixed form span.close {
      top: 7px;
    }
  }

  @media only screen and (min-width: 990px) {

    .subscribe-widget input[type=text] {
      width: 140px !important;
    }
  }

@media only screen and (min-width: 1180px) {
  .subscribe-widget input[type=text] {
    width: 180px !important;
  }
}
</style>
