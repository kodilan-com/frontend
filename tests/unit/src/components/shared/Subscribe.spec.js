import Subscribe from '@/components/shared/Subscribe';
import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Multiselect from 'vue-multiselect';

const localVue = createLocalVue();
const modalStub = {
  show: jest.fn(),
};

localVue.use(Vuex);
localVue.use({
  install(Vue) {
    // eslint-disable-next-line no-param-reassign
    Vue.prototype.$modal = modalStub;
  },
});

localVue.component('multiselect', Multiselect);

describe('Subscribe.vue', () => {
  let wrapper;
  let storeStub;
  let store;

  beforeEach(() => {
    storeStub = {
      actions: {
        subscribe: jest.fn(),
      },
    };
    store = new Vuex.Store(storeStub);

    wrapper = shallowMount(Subscribe, {
      store,
      localVue,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();

    wrapper.destroy();
  });

  it('should be shown if isVisible is true', () => {
    expect(wrapper.find('.subscribe-widget')
      .exists())
      .toBeTruthy();
    expect(wrapper.element)
      .toMatchSnapshot();
  });

  it('should be shown if isClosed is false and not subscribeClosed before', () => {
    localStorage.getItem.mockReturnValue(false);

    wrapper.setProps({
      fixed: true,
    });

    expect(wrapper.find('.subscribe-widget')
      .exists())
      .toBeTruthy();
  });

  it('should not be shown if isClosed is true', () => {
    wrapper.setProps({
      fixed: true,
    });

    wrapper.setData({
      isClosed: true,
    });

    expect(wrapper.find('.subscribe-widget')
      .exists())
      .toBeFalsy();
  });

  it('should not be shown if subscribeClosed before', () => {
    localStorage.getItem.mockReturnValue(true);

    wrapper.setProps({
      fixed: true,
    });

    expect(wrapper.find('.subscribe-widget')
      .exists())
      .toBeFalsy();
  });

  it('should call close method if "Bir daha gösterme :(" is clicked', async () => {
    wrapper.setProps({
      fixed: true,
    });

    await localVue.nextTick();

    wrapper.find('.close')
      .trigger('click');

    expect(wrapper.vm.isClosed)
      .toBe(true);
    expect(localStorage.setItem)
      .toBeCalledWith('subscribeClosed', true);
  });

  it('should not call subscribe action if isClosed is true', () => {
    wrapper.setData({
      isClosed: true,
    });

    wrapper.find('input[placeholder="Email"]')
      .trigger('keyup.enter');

    expect(storeStub.actions.subscribe)
      .not
      .toBeCalled();
  });

  it('should call subscribe action showDialog method with success message if isClosed is false', (done) => {
    const closeStub = jest.fn();

    storeStub.actions.subscribe.mockResolvedValue({});

    wrapper.setData({
      isClosed: false,
      name: 'testName',
      email: 'testMail',
    });
    wrapper.setMethods({
      close: closeStub,
    });

    wrapper.find('input[placeholder="Email"]')
      .trigger('keyup.enter');

    expect(storeStub.actions.subscribe.mock.calls[0][1])
      .toEqual({
        email: 'testMail',
        frequency: 'weekly',
        name: 'testName',
      });

    setTimeout(() => {
      expect(closeStub)
        .toBeCalled();
      expect(modalStub.show)
        .toBeCalledWith('dialog', {
          buttons: [{ title: 'Kapat' }],
          text: undefined,
          title: 'Email listesine kaydınız gerçekleştirildi.',
        });

      done();
    });
  });

  it('should call subscribe action and showDialog method with error message if isClosed is false ', (done) => {
    const closeStub = jest.fn();
    const error = {
      response: {
        data: {
          errors: [['testError']],
        },
      },
    };

    storeStub.actions.subscribe.mockRejectedValue(error);

    wrapper.setData({
      isClosed: false,
      name: 'testName',
      email: 'testMail',
    });
    wrapper.setMethods({
      close: closeStub,
    });

    wrapper.find('input[placeholder="Email"]')
      .trigger('keyup.enter');

    expect(storeStub.actions.subscribe.mock.calls[0][1])
      .toEqual({
        email: 'testMail',
        frequency: 'weekly',
        name: 'testName',
      });

    setTimeout(() => {
      expect(closeStub)
        .not
        .toBeCalled();
      expect(modalStub.show)
        .toBeCalledWith('dialog', {
          buttons: [{ title: 'Kapat' }],
          text: '<ul><li>testError</li></ul>',
          title: 'Hata: Kaydınız gerçekleştirilemedi.',
        });

      done();
    });
  });

  it('should call subscribe action and showDialog method with empty error list message if isClosed is false ', (done) => {
    const closeStub = jest.fn();
    const error = {
      response: {
        data: {},
      },
    };

    storeStub.actions.subscribe.mockRejectedValue(error);

    wrapper.setData({
      isClosed: false,
      name: 'testName',
      email: 'testMail',
    });
    wrapper.setMethods({
      close: closeStub,
    });

    wrapper.find('.button')
      .trigger('click');

    expect(storeStub.actions.subscribe.mock.calls[0][1])
      .toEqual({
        email: 'testMail',
        frequency: 'weekly',
        name: 'testName',
      });

    setTimeout(() => {
      expect(closeStub)
        .not
        .toBeCalled();
      expect(modalStub.show)
        .toBeCalledWith('dialog', {
          buttons: [{ title: 'Kapat' }],
          text: '<ul></ul>',
          title: 'Hata: Kaydınız gerçekleştirilemedi.',
        });

      done();
    });
  });
});
