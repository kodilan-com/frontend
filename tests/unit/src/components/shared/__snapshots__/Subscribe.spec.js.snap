// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscribe.vue should be shown if isVisible is true 1`] = `
<div
  class="subscribe-widget"
>
  <span>
    Güncel iş ilanlarını
  </span>
   
  <div
    class="subscribe--select"
  >
    <multiselect-stub
      allowempty="true"
      blockkeys=""
      clearonselect="true"
      closeonselect="true"
      customlabel="function(t,e){return i(t)?\\"\\":e?t[e]:t}"
      deselectgrouplabel="Press enter to deselect group"
      deselectlabel="Press enter to remove"
      internalsearch="true"
      label="text"
      limit="99999"
      limittext="function(t){return\\"and \\".concat(t,\\" more\\")}"
      maxheight="300"
      name=""
      opendirection=""
      optionheight="40"
      options="[object Object],[object Object]"
      optionslimit="1000"
      placeholder="Select option"
      selectedlabel="Selected"
      selectgrouplabel="Press enter to select group"
      selectlabel="Press enter to select"
      showlabels="true"
      shownooptions="true"
      shownoresults="true"
      showpointer="true"
      tabindex="0"
      tagplaceholder="Press enter to create a tag"
      tagposition="top"
      value="[object Object]"
    />
  </div>
   
  <span>
    olarak almak için
  </span>
   
  <input
    class="username"
    placeholder="İsim soyisim"
    type="text"
  />
   
  <input
    placeholder="Email"
    type="text"
  />
   
  <button
    class="button"
  >
    
    Abone ol!
  
  </button>
   
  <!---->
</div>
`;
